src/machi_lifecycle_mgr.erl

```diff
@@ -0,0 +1,1079 @@
+    {ok, State}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+finish_init(S) ->
+    %% machi_flu_sup will start all FLUs that have a valid definition
+    %% file.  That supervisor's structure + OTP supervisor behavior
+    %% guarantees that those FLUs should now be running.
+    %% (TODO: Unless they absolutely cannot keep running and the
+    %% supervisor has given up and terminated them.)
+    RunningFLUs = get_local_running_flus(),
+    RunningFLU_Epochs = get_latest_public_epochs(RunningFLUs),
+    RunningFLUs_at_zero = [FLU || {FLU, 0} <- RunningFLU_Epochs],
+    lager:info("Running FLUs: ~p\n", [RunningFLUs]),
+    lager:info("Running FLUs at epoch 0: ~p\n", [RunningFLUs_at_zero]),
+    ChainDefs = get_initial_chains(),
+    perhaps_bootstrap_chains(ChainDefs, RunningFLUs_at_zero, RunningFLUs),
+    S#state{flus=RunningFLUs, chains=ChainDefs}.
+
+get_local_running_flus() ->
+    [Name || {Name,_,_,_} <- supervisor:which_children(machi_flu_sup)].
+
+get_latest_public_epochs(FLUs) ->
+    [begin
+         PS = machi_flu1:make_projection_server_regname(FLU),
+         {ok, {Epoch, _CSum}} = machi_projection_store:get_latest_epochid(
+                                  PS, public),
+         {FLU, Epoch}
+     end || FLU <- FLUs].
+
+get_initial_chains() ->
+    ConfigDir = get_chain_config_dir(),
+    CDefs = [CDef || {_File, CDef} <- machi_flu_sup:load_rc_d_files_from_dir(
+                                        ConfigDir)],
+    sanitize_chain_def_records(CDefs).
+
+sanitize_chain_def_records(Ps) ->
+    {Sane, _} = lists:foldl(fun sanitize_chain_def_rec/2, {[], gb_trees:empty()}, Ps),
+    Sane.
+
+sanitize_chain_def_rec(Whole, {Acc, D}) ->
+    try
+        #chain_def_v1{name=Name,
+                      mode=Mode,
+                      full=Full,
+                      witnesses=Witnesses} = Whole,
+        {true, 10} = {is_atom(Name), 10},
+        NameK = {name, Name},
+        {none, 20} = {gb_trees:lookup(NameK, D), 20},
+        {true, 30} = {(Mode == ap_mode orelse Mode == cp_mode), 30},
+        IsPSrvr = fun(X) when is_record(X, p_srvr) -> true;
+                     (_)                           -> false
+                  end,
+        {true, 40} = {lists:all(IsPSrvr, Full), 40},
+        {true, 50} = {lists:all(IsPSrvr, Witnesses), 50},
+
+        %% All is sane enough.
+        D2 = gb_trees:enter(NameK, Name, D),
+        {[Whole|Acc], D2}
+    catch X:Y ->
+            lager:error("~s: Bad chain_def record (~w ~w), skipping: ~P\n",
+                        [?MODULE, X, Y, Whole, 15]),
+            {Acc, D}
+    end.
+
+perhaps_bootstrap_chains([], LocalFLUs_at_zero, LocalFLUs) ->
+    if LocalFLUs == [] ->
+            ok;
+       LocalFLUs_at_zero == [] ->
+            ok;
+       true ->
+            lager:warning("The following FLUs are defined but are not also "
+                          "members of a defined chain: ~w\n",
+                          [LocalFLUs_at_zero])
```

Comment:
Glad you thought of this case where a FLU is configured but not included in a chain. :cake: